build/apple/contentScope.js

```diff
@@ -2232,21 +2330,51 @@
       const { sessionKey, site } = args;
       const domainKey = site.domain;
       const featureName = 'fingerprinting-canvas';
+      const supportsWebGl = getFeatureSettingEnabled(featureName, args, 'webGl');
 
       const unsafeCanvases = new WeakSet();
+      const canvasContexts = new WeakMap();
       const canvasCache = new WeakMap();
 
+      /**
+       * Clear cache as canvas has changed
+       * @param {HTMLCanvasElement} canvas
+       */
       function clearCache (canvas) {
-          // Clear cache as canvas has changed
           canvasCache.delete(canvas);
       }
 
+      /**
+       * @param {HTMLCanvasElement} canvas
+       */
+      function treatAsUnsafe (canvas) {
+          unsafeCanvases.add(canvas);
+          clearCache(canvas);
+      }
+
+      const proxy = new DDGProxy(featureName, HTMLCanvasElement.prototype, 'getContext', {
+          apply (target, thisArg, args) {
+              const context = DDGReflect.apply(target, thisArg, args);
+              try {
+                  canvasContexts.set(thisArg, context);
+              } catch {
+              }
+              return context
+          }
+      });
+      proxy.overload();
+
       // Known data methods
       const safeMethods = ['putImageData', 'drawImage'];
       for (const methodName of safeMethods) {
           const safeMethodProxy = new DDGProxy(featureName, CanvasRenderingContext2D.prototype, methodName, {
               apply (target, thisArg, args) {
-                  clearCache(thisArg.canvas);
+                  // Don't apply escape hatch for canvases
+                  if (methodName === 'drawImage' && args[0] && args[0] instanceof HTMLCanvasElement) {
```

Comment:
Why only drawImage?

I haven't tested this but reading the [ImageData MDN](https://developer.mozilla.org/en-US/docs/Web/API/ImageData) page it seems like a site could do:

`cleanCanvas.putImageData(dirtyCanvas.getImageData(0, 0, dirtyCanvas.width, dirtyCanvas.height), 0, 0);`

Reply:
It might not be possible to only flag suspicious calls since the type of the first argument to putImageData will always be ImageData.

Reply:
So the getImageData in this example should be hooked and randomised. The raw canvas object circumvented the protections.